### PR TITLE
Start metrics listener after specialization 

### PR DESCRIPTION
**Pull request by liliankasem**
### Issue describing the changes in this PR

A bug was found where it appears that metrics are being captured pre-specialization and being persisted, leading to an inflated started invocation count e.g:

`'AppFailureCount':0,'ActiveInvocationCount':0,'StartedInvocationCount':123` 

With this change, we are waiting for specialization to happen before we collect metrics. I am deploying these changes to test them out on a legion stamp.

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

### Additional information

Additional PR information

### CODE REVIEW

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs [hunk 1]
```diff
@@ -1,30 +1,48 @@
...
 {
     public class HostMetricsProvider : IHostMetricsProvider, IDisposable
     {
         private readonly MeterListener _meterListener;
         private readonly IServiceProvider _serviceProvider;
+        private readonly IOptionsMonitor<StandbyOptions> _standbyOptions;
+        private readonly ILogger<HostMetricsProvider> _logger;
+        private readonly object _lock = new object();
+
         private ConcurrentDictionary<string, long> _metricsCache = new();
```
Comment (kshyju) [inline src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs:26]: Does `_metricCache` needs to be  concurrent dictionary? Looks like all operations on that is done within a `lock` block.

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs [hunk 2]
```diff
@@ -46,22 +64,42 @@ public void Start()
...
+                Start();
+            }
         }
 
         internal void OnMeasurementRecordedLong(
             Instrument instrument,
             long measurement,
-            ReadOnlySpan<KeyValuePair<string, object?>> tags,
-            object? state)
+            ReadOnlySpan<KeyValuePair<string, object>> tags,
```
Comment (kshyju) [inline src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs:87]: Should this method be private?

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs [hunk 3]
```diff
@@ -86,43 +124,42 @@ private void AddOrUpdateMetricsCache(string key, long value)
...
+                metrics = new Dictionary<string, long>(_metricsCache);
+                _metricsCache.Clear();
             }
+
+            return metrics;
         }
 
         public bool HasMetrics()
         {
             return _metricsCache.Count > 0;
```
Comment (kshyju) [inline src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs]: fyi - this type has a `IsEmpty` property as well.

File: src/WebJobs.Script.WebHost/Metrics/IHostMetricsProvider.cs [hunk 4]
```diff
@@ -24,6 +24,11 @@ public interface IHostMetricsProvider
         /// </summary>
         public string FunctionGroup { get; }
 
+        /// <summary>
+        /// Initializes the provider and starts collecting metrics.
+        /// </summary>
+        public void Start();
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/IHostMetricsProvider.cs]: Should this change be reverted? In your initial impl. I saw external calls to this, but in the latest it's internal calls again? The metrics provider is starting itself again on specialization.

**Review by kshyju — APPROVED** [review]
One question about the need of concurrent dictionary & lock

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs (hunk 1 above)
Comment (mathewc) [reply src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs:26]: Furthermore, there is code below that is recreating it as a regular dictionary so yes this should be fixed

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs [hunk 5]
```diff
@@ -86,43 +124,42 @@ private void AddOrUpdateMetricsCache(string key, long value)
...
+                AddOrUpdateMetricsCache(HostMetrics.ActiveInvocationCount, functionActivityStatus.OutstandingInvocations);
             }
-            finally
+
+            IReadOnlyDictionary<string, long> metrics;
+
+            lock (_lock)
             {
-                PurgeMetricsCache();
+                metrics = new Dictionary<string, long>(_metricsCache);
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs]: Initially defined as concurrent, but here not concurrent

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs [hunk 6]
```diff
@@ -46,22 +64,42 @@ public void Start()
...
+                return;
+            }
+
             if (instrument == null)
             {
                 throw new ArgumentNullException(nameof(instrument));
             }
 
-            AddOrUpdateMetricsCache(instrument.Name, measurement);
+            lock (_lock)
```
Comment (mathewc) [inline src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs]: Confused why some calls to AddOrUpdate are done inside a lock but others aren't. If locking is really needed, can't that be pushed down into AddOrUpdate? Or perhaps just use a concurrent dictionary like you were before?

File: src/WebJobs.Script.WebHost/Metrics/IHostMetricsProvider.cs (hunk 4 above)
Comment (liliankasem) [reply src/WebJobs.Script.WebHost/Metrics/IHostMetricsProvider.cs]: Good catch, will remove

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs (hunk 6 above)
Comment (liliankasem) [reply src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs]: Yeah I think I confused everything messing with this; I think I should stick on a concurrent dictionary as before.

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs (hunk 5 above)
Comment (liliankasem) [reply src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs]: Edit: woopies, I see how now we're replacing the concurrent one with a plain old dictionary - thanks for catching

File: src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs (hunk 1 above)
Comment (liliankasem) [reply src/WebJobs.Script.WebHost/Metrics/HostMetricsProvider.cs:26]: Switching back to ConcurrentDictionary without a lock; Friday evening coding did me bad 😅

**Review by mathewc — APPROVED** [review]

**Pull request merged**